embedded/appendable/remoteapp/remote_app.go

```diff
@@ -187,58 +182,52 @@ func (r *RemoteStorageAppendable) uploadChunk(chunkID int64, dontRemoveFile bool
 		defer metricsUploadFinished.Inc()
 
 		var newApp appendable.Appendable
-		cp := r.chunkedProcess(ctx)
+		cp := chain.New().WithContext(ctx).WithBackoff(r.backoff)
 
 		// Chunk data was already flushed when changing the active appendable
 
 		// Update size stats after flush
-		cp.Step(func() error {
+		cp.Next(func() error {
 			fi, err := os.Stat(fileName)
 			if err == nil {
 				r.mutex.Lock()
 				r.chunkInfos[chunkID].storageSize = fi.Size()
 				r.mutex.Unlock()
 			}
 			return nil
-		})
-
-		// Upload the chunk
-		cp.RetryableStep(func(retries int, delay time.Duration) (bool, error) {
+		}).NextWithRetry(func() error { // Upload the chunk
 			defer prometheus.NewTimer(metricsUploadTime).ObserveDuration()
 			err := r.rStorage.Put(ctx, r.remotePath+appName, fileName)
 			if err == nil {
-				return false, nil
+				return nil
 			}
 
 			metricsUploadRetried.Inc()
-			return true, nil
-		})
-
-		// Wait for the chunk to become ready
-		cp.RetryableStep(func(retries int, delay time.Duration) (bool, error) {
+			return err
+		}).NextWithRetry(func() error { // Wait for the chunk to become ready
 			exists, err := r.rStorage.Exists(ctx, r.remotePath+appName)
 			if err == nil && exists {
-				return false, nil
+				return nil
 			}
 
 			metricsUploadRetried.Inc()
-			return true, nil
-		})
 
-		// Open new appendable from the remote storage
-		cp.RetryableStep(func(retries int, delay time.Duration) (bool, error) {
+			// Retry since there may be some asynchronous processing on the remote
+			// storage and we want to ensure that the new object is already visible.
+			if err == nil && !exists {
```

Comment:
@byo does this step fix this comment?

https://github.com/codenotary/immudb/pull/1402/files#r984674900

Reply:
Looks like it solves this case. The code may even be simplified (to expose the happy path better):

```go
		}).NextWithRetry(func() error { // Wait for the chunk to become ready
			exists, err := r.rStorage.Exists(ctx, r.remotePath+appName)
			if err != nil {
				metricsUploadRetried.Inc()
				return err
			}

			if !exists {
				// Retry since there may be some asynchronous processing on the remote
				// storage and we want to ensure that the new object is already visible.
				return retry.ErrRetry
			}

			return nil
		}).NextWithRetry(func() error { // Open new appendable from the remote storage
```

The only difference to the current code is the `metricsUploadRetried.Inc()` - but now I think it does not need to be updated when the remote storage is still processing the upload and not yet exposes the uploaded blob. We're interested in other errors though.

Reply:
Ack, thanks for the pointer, I'll fix this